lib/wasm/WebAssemblyGenerator.js

```diff
@@ -229,35 +239,28 @@ const addInitFunction = ({
 	// Export section
 	const moduleExport = t.moduleExport(initFuncId.value, "Func", nextFuncIndex);
 
-	return add(bin, [func, moduleExport, funcindex, functype]);
+	return addWithAST(ast, bin, [func, moduleExport, funcindex, functype]);
 };
 
 class WebAssemblyGenerator extends Generator {
 	generate(module) {
+		const ast = WebAssemblyParser.getAst(module);
 		const bin = module.originalSource().source();
 
-		// FIXME(sven): this module is parsed twice, we could preserve the AST
-		// from wasm/WebAssemblyParser.js
-		const ast = decode(bin, {
-			ignoreDataSection: true,
-			ignoreCodeSection: true
-		});
```

Comment:
As far I understand the behavior of `edit/addWithAst` these methods modify the AST while processing. To ensure correctness we would need to clone the AST that this position to ensure we don't modify the original AST. Otherwise we get into problems when Generator is called twice for a single Parser call (which could happen in watch mode).

Or we keep the decoding call here.

Reply:
That's a good point. The AST can be quite big and I'd like to avoid copying it to much (like for each edit/add).

Decoding it here too sounds good to me, we are still avoiding 4 more.

Reply:
I'll finish that for you. This also means we could remove the AST passing from parser to generator.

Reply:
Yes, thanks.